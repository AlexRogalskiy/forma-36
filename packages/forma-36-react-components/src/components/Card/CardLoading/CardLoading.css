@import 'settings/colors';
@import 'settings/typography';
@import 'settings/transitions';

.CardLoading {
  display: block;
  position: relative;
  box-sizing: border-box;
  font-family: var(--font-stack-primary);
  font-size: var(--font-size-m);
  line-height: var(--line-height-default);
  color: var(--color-text-dark);
}

.CardLoading--is-loading {
  height: 80px; /* Provides a default height as card only contains a Spinner */
  overflow: hidden;
}

.CardLoading__spinner {
  position: absolute;
  top: calc(50% - 10px); /* 10px for offsetting spinner by its height */
  left: calc(50% - 10px); /* 10px for offsetting spinner by its width */
}

.CardLoading__spinner--enter {
  opacity: 0;
}

.CardLoading__spinner--enter-active {
  opacity: 1;
  transition: opacity var(--transition-duration-short)
    var(--transition-easing-default);
}

.CardLoading__spinner--exit {
  opacity: 1;
}

.CardLoading__spinner--exit-active {
  opacity: 0;
  transition: opacity var(--transition-duration-short)
    var(--transition-easing-default);
}

.CardLoading__wrapper--enter {
  opacity: 0;
}

.CardLoading__wrapper--enter-active {
  transition: opacity var(--transition-duration-short)
    var(--transition-easing-default);
  opacity: 1;
}

.CardLoading__wrapper--exit {
  opacity: 1;
}

.CardLoading__wrapper--exit-active {
  opacity: 0;
  transition: opacity var(--transition-duration-short)
    var(--transition-easing-default);
}
