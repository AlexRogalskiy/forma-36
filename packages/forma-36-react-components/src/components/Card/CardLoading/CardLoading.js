import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import { CSSTransition } from 'react-transition-group';
import AnimateHeight from 'react-animate-height';
import Card from '../Card';
import Spinner from '../../Spinner';
import styles from './CardLoading.css';

class CardLoading extends React.Component {
  static propTypes = {
    extraClassNames: PropTypes.string,
    title: PropTypes.string,
    children: PropTypes.node.isRequired,
    testId: PropTypes.string,
    isLoading: PropTypes.bool,
    onClick: PropTypes.func,
  };

  static defaultProps = {
    title: 'Untitled',
    testId: 'cf-ui-embedded-card',
    isLoading: undefined,
    onClick: undefined,
    extraClassNames: undefined,
  };

  state = {
    selected: false,
  };

  handleClick = event => {
    if (!this.props.onClick) return;

    this.setState({ selected: !this.state.selected });
    this.props.onClick(event);
  };

  render() {
    const {
      extraClassNames,
      children,
      testId,
      isLoading,
      ...otherProps
    } = this.props;

    const classNames = cn(styles.CardLoading, extraClassNames, {
      [styles['CardLoading--is-loading']]: isLoading,
    });

    return (
      <AnimateHeight
        height={isLoading ? 120 : 'auto'}
        duration={100}
        easing="ease-in-out"
      >
        <Card
          extraClassNames={classNames}
          testId={testId}
          onClick={!isLoading && (event => this.handleClick(event))}
          selected={this.state.selected}
          {...otherProps}
        >
          <CSSTransition
            timeout={100}
            in={isLoading}
            classNames={{
              enter: styles['CardLoading--enter'],
              enterActive: styles['CardLoading__spinner--enter-active'],
              exit: styles['CardLoading__spinner--exit'],
              exitActive: styles['CardLoading__spinner--exit-active'],
            }}
            mountOnEnter
            unmountOnExit
          >
            <Spinner extraClassNames={styles.CardLoading__spinner} />
          </CSSTransition>
          <CSSTransition
            timeout={100}
            in={!isLoading}
            classNames={{
              enter: styles['CardLoading__wrapper--enter'],
              enterActive: styles['CardLoading__wrapper--enter-active'],
              exit: styles['CardLoading__wrapper--exit'],
              exitActive: styles['CardLoading__wrapper--exit-active'],
            }}
            mountOnEnter
            unmountOnExit
          >
            <React.Fragment>{children}</React.Fragment>
          </CSSTransition>
        </Card>
      </AnimateHeight>
    );
  }
}

export default CardLoading;
