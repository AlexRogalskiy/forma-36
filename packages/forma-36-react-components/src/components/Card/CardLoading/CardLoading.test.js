import React from 'react';
import { shallow, mount } from 'enzyme';
import { axe } from 'jest-axe';
import CardLoading from './CardLoading';

it('renders the component', () => {
  const output = shallow(<CardLoading>CardLoading</CardLoading>);

  expect(output).toMatchSnapshot();
});

it('renders the component with an additional class name', () => {
  const output = shallow(
    <CardLoading extraClassNames="my-extra-class">CardLoading</CardLoading>,
  );

  expect(output).toMatchSnapshot();
});

it('renders a loading spinner when passed a "loading" prop', () => {
  const output = mount(<CardLoading isLoading>Asset Card</CardLoading>);

  expect(output.find('Spinner').length).toEqual(1);

  expect(output).toMatchSnapshot();
});

it('does not render a loading spinner when not passed a "loading" prop', () => {
  const output = mount(<CardLoading>Asset Card</CardLoading>);

  expect(output.find('Spinner').length).toEqual(0);

  expect(output).toMatchSnapshot();
});

it('has no a11y issues', async () => {
  const output = mount(<CardLoading>CardLoading</CardLoading>).html();
  const results = await axe(output);

  expect(results).toHaveNoViolations();
});
