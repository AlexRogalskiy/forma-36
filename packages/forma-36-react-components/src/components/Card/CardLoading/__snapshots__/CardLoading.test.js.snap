// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not render a loading spinner when not passed a "loading" prop 1`] = `
<CardLoading
  testId="cf-ui-embedded-card"
  title="Untitled"
>
  <AnimateHeight
    animateOpacity={false}
    animationStateClasses={
      Object {
        "animating": "rah-animating",
        "animatingDown": "rah-animating--down",
        "animatingToHeightAuto": "rah-animating--to-height-auto",
        "animatingToHeightSpecific": "rah-animating--to-height-specific",
        "animatingToHeightZero": "rah-animating--to-height-zero",
        "animatingUp": "rah-animating--up",
        "static": "rah-static",
        "staticHeightAuto": "rah-static--height-auto",
        "staticHeightSpecific": "rah-static--height-specific",
        "staticHeightZero": "rah-static--height-zero",
      }
    }
    applyInlineTransitions={true}
    delay={0}
    duration={100}
    easing="ease-in-out"
    height="auto"
    style={Object {}}
  >
    <div
      aria-hidden={false}
      className="rah-static rah-static--height-auto"
      style={
        Object {
          "height": "auto",
          "overflow": "visible",
        }
      }
    >
      <div
        style={Object {}}
      >
        <Card
          extraClassNames="CardLoading"
          padding="default"
          selected={false}
          testId="cf-ui-embedded-card"
          title="Untitled"
        >
          <div
            className="Card CardLoading Card--padding-default"
            data-test-id="cf-ui-embedded-card"
            title="Untitled"
          >
            <CSSTransition
              classNames={
                Object {
                  "enter": "CardLoading--enter",
                  "enterActive": "CardLoading__spinner--enter-active",
                  "exit": "CardLoading__spinner--exit",
                  "exitActive": "CardLoading__spinner--exit-active",
                }
              }
              mountOnEnter={true}
              timeout={100}
              unmountOnExit={true}
            >
              <Transition
                appear={false}
                enter={true}
                exit={true}
                in={false}
                mountOnEnter={true}
                onEnter={[Function]}
                onEntered={[Function]}
                onEntering={[Function]}
                onExit={[Function]}
                onExited={[Function]}
                onExiting={[Function]}
                timeout={100}
                unmountOnExit={true}
              />
            </CSSTransition>
            <CSSTransition
              classNames={
                Object {
                  "enter": "CardLoading__wrapper--enter",
                  "enterActive": "CardLoading__wrapper--enter-active",
                  "exit": "CardLoading__wrapper--exit",
                  "exitActive": "CardLoading__wrapper--exit-active",
                }
              }
              in={true}
              mountOnEnter={true}
              timeout={100}
              unmountOnExit={true}
            >
              <Transition
                appear={false}
                enter={true}
                exit={true}
                in={true}
                mountOnEnter={true}
                onEnter={[Function]}
                onEntered={[Function]}
                onEntering={[Function]}
                onExit={[Function]}
                onExited={[Function]}
                onExiting={[Function]}
                timeout={100}
                unmountOnExit={true}
              >
                Asset Card
              </Transition>
            </CSSTransition>
          </div>
        </Card>
      </div>
    </div>
  </AnimateHeight>
</CardLoading>
`;

exports[`renders a loading spinner when passed a "loading" prop 1`] = `
<CardLoading
  isLoading={true}
  testId="cf-ui-embedded-card"
  title="Untitled"
>
  <AnimateHeight
    animateOpacity={false}
    animationStateClasses={
      Object {
        "animating": "rah-animating",
        "animatingDown": "rah-animating--down",
        "animatingToHeightAuto": "rah-animating--to-height-auto",
        "animatingToHeightSpecific": "rah-animating--to-height-specific",
        "animatingToHeightZero": "rah-animating--to-height-zero",
        "animatingUp": "rah-animating--up",
        "static": "rah-static",
        "staticHeightAuto": "rah-static--height-auto",
        "staticHeightSpecific": "rah-static--height-specific",
        "staticHeightZero": "rah-static--height-zero",
      }
    }
    applyInlineTransitions={true}
    delay={0}
    duration={100}
    easing="ease-in-out"
    height={120}
    style={Object {}}
  >
    <div
      aria-hidden={false}
      className="rah-static rah-static--height-specific"
      style={
        Object {
          "height": 120,
          "overflow": "hidden",
        }
      }
    >
      <div
        style={Object {}}
      >
        <Card
          extraClassNames="CardLoading CardLoading--is-loading"
          padding="default"
          selected={false}
          testId="cf-ui-embedded-card"
          title="Untitled"
        >
          <div
            className="Card CardLoading CardLoading--is-loading Card--padding-default"
            data-test-id="cf-ui-embedded-card"
            title="Untitled"
          >
            <CSSTransition
              classNames={
                Object {
                  "enter": "CardLoading--enter",
                  "enterActive": "CardLoading__spinner--enter-active",
                  "exit": "CardLoading__spinner--exit",
                  "exitActive": "CardLoading__spinner--exit-active",
                }
              }
              in={true}
              mountOnEnter={true}
              timeout={100}
              unmountOnExit={true}
            >
              <Transition
                appear={false}
                enter={true}
                exit={true}
                in={true}
                mountOnEnter={true}
                onEnter={[Function]}
                onEntered={[Function]}
                onEntering={[Function]}
                onExit={[Function]}
                onExited={[Function]}
                onExiting={[Function]}
                timeout={100}
                unmountOnExit={true}
              >
                <Spinner
                  color="default"
                  extraClassNames="CardLoading__spinner"
                  size="default"
                  testId="cf-ui-spinner"
                >
                  <svg
                    className="Spinner CardLoading__spinner Spinner--default"
                    data-test-id="cf-ui-spinner"
                    style={null}
                    viewBox="0 0 20 20"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <title>
                      Loadingâ€¦
                    </title>
                    <path
                      d="M2,10a8,8,0,0,1,8-8V0a10,10,0,0,0,0,20V18A8,8,0,0,1,2,10Z"
                    />
                    <path
                      d="M10,0V2a8,8,0,0,1,0,16v2A10,10,0,0,0,10,0Z"
                      style={
                        Object {
                          "opacity": 0.4,
                        }
                      }
                    />
                  </svg>
                </Spinner>
              </Transition>
            </CSSTransition>
            <CSSTransition
              classNames={
                Object {
                  "enter": "CardLoading__wrapper--enter",
                  "enterActive": "CardLoading__wrapper--enter-active",
                  "exit": "CardLoading__wrapper--exit",
                  "exitActive": "CardLoading__wrapper--exit-active",
                }
              }
              in={false}
              mountOnEnter={true}
              timeout={100}
              unmountOnExit={true}
            >
              <Transition
                appear={false}
                enter={true}
                exit={true}
                in={false}
                mountOnEnter={true}
                onEnter={[Function]}
                onEntered={[Function]}
                onEntering={[Function]}
                onExit={[Function]}
                onExited={[Function]}
                onExiting={[Function]}
                timeout={100}
                unmountOnExit={true}
              />
            </CSSTransition>
          </div>
        </Card>
      </div>
    </div>
  </AnimateHeight>
</CardLoading>
`;

exports[`renders the component 1`] = `
<AnimateHeight
  animateOpacity={false}
  animationStateClasses={
    Object {
      "animating": "rah-animating",
      "animatingDown": "rah-animating--down",
      "animatingToHeightAuto": "rah-animating--to-height-auto",
      "animatingToHeightSpecific": "rah-animating--to-height-specific",
      "animatingToHeightZero": "rah-animating--to-height-zero",
      "animatingUp": "rah-animating--up",
      "static": "rah-static",
      "staticHeightAuto": "rah-static--height-auto",
      "staticHeightSpecific": "rah-static--height-specific",
      "staticHeightZero": "rah-static--height-zero",
    }
  }
  applyInlineTransitions={true}
  delay={0}
  duration={100}
  easing="ease-in-out"
  height="auto"
  style={Object {}}
>
  <Card
    extraClassNames="CardLoading"
    padding="default"
    selected={false}
    testId="cf-ui-embedded-card"
    title="Untitled"
  >
    <CSSTransition
      classNames={
        Object {
          "enter": "CardLoading--enter",
          "enterActive": "CardLoading__spinner--enter-active",
          "exit": "CardLoading__spinner--exit",
          "exitActive": "CardLoading__spinner--exit-active",
        }
      }
      mountOnEnter={true}
      timeout={100}
      unmountOnExit={true}
    >
      <Spinner
        color="default"
        extraClassNames="CardLoading__spinner"
        size="default"
        testId="cf-ui-spinner"
      />
    </CSSTransition>
    <CSSTransition
      classNames={
        Object {
          "enter": "CardLoading__wrapper--enter",
          "enterActive": "CardLoading__wrapper--enter-active",
          "exit": "CardLoading__wrapper--exit",
          "exitActive": "CardLoading__wrapper--exit-active",
        }
      }
      in={true}
      mountOnEnter={true}
      timeout={100}
      unmountOnExit={true}
    >
      CardLoading
    </CSSTransition>
  </Card>
</AnimateHeight>
`;

exports[`renders the component with an additional class name 1`] = `
<AnimateHeight
  animateOpacity={false}
  animationStateClasses={
    Object {
      "animating": "rah-animating",
      "animatingDown": "rah-animating--down",
      "animatingToHeightAuto": "rah-animating--to-height-auto",
      "animatingToHeightSpecific": "rah-animating--to-height-specific",
      "animatingToHeightZero": "rah-animating--to-height-zero",
      "animatingUp": "rah-animating--up",
      "static": "rah-static",
      "staticHeightAuto": "rah-static--height-auto",
      "staticHeightSpecific": "rah-static--height-specific",
      "staticHeightZero": "rah-static--height-zero",
    }
  }
  applyInlineTransitions={true}
  delay={0}
  duration={100}
  easing="ease-in-out"
  height="auto"
  style={Object {}}
>
  <Card
    extraClassNames="CardLoading my-extra-class"
    padding="default"
    selected={false}
    testId="cf-ui-embedded-card"
    title="Untitled"
  >
    <CSSTransition
      classNames={
        Object {
          "enter": "CardLoading--enter",
          "enterActive": "CardLoading__spinner--enter-active",
          "exit": "CardLoading__spinner--exit",
          "exitActive": "CardLoading__spinner--exit-active",
        }
      }
      mountOnEnter={true}
      timeout={100}
      unmountOnExit={true}
    >
      <Spinner
        color="default"
        extraClassNames="CardLoading__spinner"
        size="default"
        testId="cf-ui-spinner"
      />
    </CSSTransition>
    <CSSTransition
      classNames={
        Object {
          "enter": "CardLoading__wrapper--enter",
          "enterActive": "CardLoading__wrapper--enter-active",
          "exit": "CardLoading__wrapper--exit",
          "exitActive": "CardLoading__wrapper--exit-active",
        }
      }
      in={true}
      mountOnEnter={true}
      timeout={100}
      unmountOnExit={true}
    >
      CardLoading
    </CSSTransition>
  </Card>
</AnimateHeight>
`;
